In [1]:
%load_ext autoreload
%autoreload 2

import pandas as pd
import numpy as np
import scipy.stats
import statistics as st
from sklearn.metrics import confusion_matrix

import datetime

import matplotlib.pylab as plt
from matplotlib_venn import venn3
from venn import venn
%matplotlib inline

from utils import prepare_df

In [3]:
col = ['period_end_date', 'delivery_type_id', 'predict_automatch', 'country_id_n']
df = prepare_df(droprows=col)
df.head()

period_end_date           translated_when  if_data_corrected  \
2 2020-09-20 00:00:00+00:00 2020-09-23 12:27:51+00:00                  0   
3 2020-09-20 00:00:00+00:00 2020-09-23 12:26:41+00:00                  0   
4 2020-09-20 00:00:00+00:00 2020-09-22 09:40:36+00:00                  0   
5 2020-09-20 00:00:00+00:00 2020-09-25 15:31:09+00:00                  0   
6 2020-09-20 00:00:00+00:00 2020-09-22 11:18:00+00:00                  0   

   prod_gr_id  country_id_n  delivery_type_id  freq_id  retailer_id  brand_id  \
2         426         121.0           42730.0        2           30        37   
3         426         121.0           42730.0        2           92       203   
4         426         121.0           50128.0        2            9        31   
5         426         121.0           19105.0        2          145       205   
6         427         108.0           47990.0        1          172       237   

   predict_automatch  class_acctual  
2                1.0              1  
3                1.0              1  
4                1.0              1  
5                1.0              1  
6                1.0              1

In [34]:
dec1st = datetime.date(year=2020, month=12, day=1)
df1 = df[df['translated_when'].dt.date <= dec1st]

In [35]:
df2 = df[df['translated_when'].dt.date > dec1st]

In [36]:
dfn = prepare_df()
dfn.head()

period_end_date           translated_when  if_data_corrected  \
0                       NaT 2020-10-15 06:58:28+00:00                  0   
1                       NaT 2020-10-15 06:38:34+00:00                  0   
2 2020-09-20 00:00:00+00:00 2020-09-23 12:27:51+00:00                  0   
3 2020-09-20 00:00:00+00:00 2020-09-23 12:26:41+00:00                  0   
4 2020-09-20 00:00:00+00:00 2020-09-22 09:40:36+00:00                  0   

   prod_gr_id  country_id_n  delivery_type_id  freq_id  retailer_id  brand_id  \
0         426         121.0               NaN        2           96       111   
1         426           NaN               NaN        2           92        95   
2         426         121.0           42730.0        2           30        37   
3         426         121.0           42730.0        2           92       203   
4         426         121.0           50128.0        2            9        31   

   predict_automatch  class_acctual  
0                1.0              1  
1                0.0              1  
2                1.0              1  
3                1.0              1  
4                1.0              1

In [37]:
dec1st = datetime.date(year=2020, month=12, day=1)
dfn1 = dfn[dfn['translated_when'].dt.date <= dec1st]
dfn1.head()

period_end_date           translated_when  if_data_corrected  \
0                       NaT 2020-10-15 06:58:28+00:00                  0   
1                       NaT 2020-10-15 06:38:34+00:00                  0   
2 2020-09-20 00:00:00+00:00 2020-09-23 12:27:51+00:00                  0   
3 2020-09-20 00:00:00+00:00 2020-09-23 12:26:41+00:00                  0   
4 2020-09-20 00:00:00+00:00 2020-09-22 09:40:36+00:00                  0   

   prod_gr_id  country_id_n  delivery_type_id  freq_id  retailer_id  brand_id  \
0         426         121.0               NaN        2           96       111   
1         426           NaN               NaN        2           92        95   
2         426         121.0           42730.0        2           30        37   
3         426         121.0           42730.0        2           92       203   
4         426         121.0           50128.0        2            9        31   

   predict_automatch  class_acctual  
0                1.0              1  
1                0.0              1  
2                1.0              1  
3                1.0              1  
4                1.0              1

## Covariance

## Bivariate analysis

In [38]:
for name, _df in zip(['df', 'df1', 'df2', 'dfn', 'dfn1'], [df, df1, df2, dfn, dfn1]):
    print(name)
    print('covariance')
    print(np.cov(_df['predict_automatch'], _df['class_acctual']))
    print('correlation')
    print(np.corrcoef(_df['predict_automatch'], _df['class_acctual']))
    if  np.sum(pd.isna(_df['predict_automatch']))==0 and np.sum(pd.isna(_df['class_acctual'])) == 0:
        r, p = scipy.stats.pearsonr(_df['predict_automatch'], _df['class_acctual'])
        print(f'r={r}, p={p}' )

    result = scipy.stats.linregress(_df['predict_automatch'], _df['class_acctual'])
    print(f'lin regression = {result}')



df
covariance
[[0.22376943 0.12590514]
 [0.12590514 0.19133339]]
correlation
[[1.         0.60848196]
 [0.60848196 1.        ]]
r=0.6084819614535691, p=0.0
lin regression = LinregressResult(slope=0.5626556801855788, intercept=0.36975753604194067, rvalue=0.6084819614535675, pvalue=0.0, stderr=0.005460798932676295, intercept_stderr=0.004443079941328524)
df1
covariance
[[0.22286916 0.12683907]
 [0.12683907 0.19103953]]
correlation
[[1.         0.61470455]
 [0.61470455 1.        ]]
r=0.6147045539848507, p=0.0
lin regression = LinregressResult(slope=0.5691189632857133, intercept=0.36451668684073774, rvalue=0.614704553984855, pvalue=0.0, stderr=0.0055604600831566835, intercept_stderr=0.004533577914731626)
df2
covariance
[[0.23964312 0.10532652]
 [0.10532652 0.19766794]]
correlation
[[1.         0.48393513]
 [0.48393513 1.        ]]
r=0.48393512615134227, p=1.0091290177519133e-48
lin regression = LinregressResult(slope=0.43951406976150376, intercept=0.4641744548286603, rvalue=0.48393512615134

## Bivariate analysis by days

In [5]:
dfcf = df[~df['country_id_n'].isin([106.0, 109.0])]
dfcf.head() 


period_end_date           translated_when  if_data_corrected  \
2 2020-09-20 00:00:00+00:00 2020-09-23 12:27:51+00:00                  0   
3 2020-09-20 00:00:00+00:00 2020-09-23 12:26:41+00:00                  0   
4 2020-09-20 00:00:00+00:00 2020-09-22 09:40:36+00:00                  0   
5 2020-09-20 00:00:00+00:00 2020-09-25 15:31:09+00:00                  0   
6 2020-09-20 00:00:00+00:00 2020-09-22 11:18:00+00:00                  0   

   prod_gr_id  country_id_n  delivery_type_id  freq_id  retailer_id  brand_id  \
2         426         121.0           42730.0        2           30        37   
3         426         121.0           42730.0        2           92       203   
4         426         121.0           50128.0        2            9        31   
5         426         121.0           19105.0        2          145       205   
6         427         108.0           47990.0        1          172       237   

   predict_automatch  class_acctual  
2                1.0              1  
3                1.0              1  
4                1.0              1  
5                1.0              1  
6                1.0              1

In [6]:
for day in sorted(dfcf['period_end_date'].unique()):
    df_day = dfcf[dfcf['period_end_date'] == day]
    print('\n', day, '\n')
    x, y = df_day['predict_automatch'], df_day['class_acctual']
    print('covariance')
    print(np.cov(x, y))
    print('correlation')
    print(np.corrcoef(x, y))

    result = scipy.stats.linregress(x, y)
    print(f'lin regression = {result}')

    con_matrix = confusion_matrix(x, y)
    print(con_matrix)


 2020-08-30 00:00:00+00:00 

covariance
[[0.18865655 0.12110154]
 [0.12110154 0.16670767]]
correlation
[[1.         0.68286672]
 [0.68286672 1.        ]]
lin regression = LinregressResult(slope=0.6419153868443949, intercept=0.3088235294117632, rvalue=0.6828667186724068, pvalue=1.4832744874774228e-38, stderr=0.04187082042272948, intercept_stderr=0.03623888644415803)
[[ 47  21]
 [ 10 193]]

 2020-09-01 00:00:00+00:00 

covariance
[[0.21885862 0.14281858]
 [0.14281858 0.19764882]]
correlation
[[1.         0.68668216]
 [0.68668216 1.        ]]
lin regression = LinregressResult(slope=0.652560901744696, intercept=0.28740157480314743, rvalue=0.6866821596281161, pvalue=7.974571261611519e-111, stderr=0.02465696646978908, intercept_stderr=0.020291580855567648)
[[181  73]
 [ 32 501]]

 2020-09-06 00:00:00+00:00 

covariance
[[0.22232332 0.14700895]
 [0.14700895 0.18641222]]
correlation
[[1.         0.72212784]
 [0.72212784 1.        ]]
lin regression = LinregressResult(slope=0.6612394605595697, 

In [46]:
df[df['period_end_date'].dt.date == datetime.date(year=2020, month=12, day=1)]

period_end_date           translated_when  if_data_corrected  \
3172 2020-12-01 00:00:00+00:00 2020-11-17 07:23:45+00:00                  1   
3967 2020-12-01 00:00:00+00:00 2020-11-20 08:34:30+00:00                  1   
5566 2020-12-01 00:00:00+00:00 2020-11-17 14:22:33+00:00                  1   

      prod_gr_id  country_id_n  delivery_type_id  freq_id  retailer_id  \
3172         426         114.0           36120.0        2           93   
3967         426         114.0           36120.0        1          120   
5566         426         114.0           36120.0        2           93   

      brand_id  predict_automatch  class_acctual  
3172        96                1.0              1  
3967       259                0.0              0  
5566        96                0.0              1

## Bivariate analysis by country

In [4]:
for country in sorted(df['country_id_n'].unique()):
    df_country = df[df['country_id_n'] == country]
    
    x, y = df_country ['predict_automatch'], df_country ['class_acctual']
    print('\n', country , x.shape, y.shape, '\n')
    if x.shape[0] and y.shape[0]:
        print('covariance')
        print(np.cov(x, y))
        print('correlation')
        print(np.corrcoef(x, y))
        con_matrix = confusion_matrix(x, y)
        print(con_matrix)
        # print(f'lin regression = {result}')
        # matrix = np.cov(x, y).round(decimals=2)
        # fig, ax = plt.subplots()
        # ax.imshow(matrix)
        # ax.grid(False)
        # ax.set_title(f'country id: {country}')
        # ax.xaxis.set(ticks=(0, 1), ticklabels=('x', 'y'))
        # ax.yaxis.set(ticks=(0, 1), ticklabels=('x', 'y'))
        # ax.set_ylim(1.5, -0.5)
        # for i in range(2):
        #     for j in range(2):
        #         ax.text(j, i, matrix[i, j], ha='center', va='center', color='w')
        # plt.show()
        # result = scipy.stats.linregress(x, y)
        # fig, ax = plt.subplots()
        # ax.set_title(f'country id: {country}')
        # ax.plot(x, y, linewidth=0, marker='s', label='Data points')
        # ax.plot(x, result.intercept + result.slope * x, label="")
        # ax.set_xlabel('x')
        # ax.set_ylabel('y')
        # ax.legend(facecolor='white')
        # plt.show()





 103.0 (877,) (877,) 

covariance
[[0.20403461 0.1479072 ]
 [0.1479072  0.18350749]]
correlation
[[1.         0.76438205]
 [0.76438205 1.        ]]
[[190  60]
 [ 22 605]]

 104.0 (688,) (688,) 

covariance
[[0.2083016  0.15430038]
 [0.15430038 0.19552275]]
correlation
[[1.         0.76457844]
 [0.76457844 1.        ]]
[[160  43]
 [ 23 462]]

 105.0 (350,) (350,) 

covariance
[[0.22025379 0.16148997]
 [0.16148997 0.19295129]]
correlation
[[1.         0.78335753]
 [0.78335753 1.        ]]
[[ 86  28]
 [  5 231]]

 106.0 (726,) (726,) 

covariance
[[ 0.24559514 -0.07200722]
 [-0.07200722  0.19924575]]
correlation
[[ 1.         -0.32551547]
 [-0.32551547  1.        ]]
[[ 61 352]
 [138 175]]

 107.0 (356,) (356,) 

covariance
[[0.22028802 0.15511948]
 [0.15511948 0.2000633 ]]
correlation
[[1.         0.73890259]
 [0.73890259 1.        ]]
[[ 87  29]
 [ 11 229]]

 108.0 (3125,) (3125,) 

covariance
[[0.21277275 0.14942812]
 [0.14942812 0.19238474]]
correlation
[[1.         0.73856549]
 [0.738

Countries bad covariance: 109, 106

## Bivariate analysis by if_data_corrected

In [7]:
for data_corr in sorted(dfcf['if_data_corrected'].unique()):
    df_data_corr = dfcf[dfcf['if_data_corrected'] == data_corr]
    print('\n', data_corr, '\n')
    x, y = df_data_corr['predict_automatch'], df_data_corr['class_acctual']
    print('covariance')
    print(np.cov(x, y))
    print('correlation')
    print(np.corrcoef(x, y))

    result = scipy.stats.linregress(x, y)
    print(f'lin regression = {result}')

    con_matrix = confusion_matrix(x, y)
    print(con_matrix)


 0 

covariance
[[0.21844342 0.14034459]
 [0.14034459 0.19177413]]
correlation
[[1.         0.68569531]
 [0.68569531 1.        ]]
lin regression = LinregressResult(slope=0.6424756966504725, intercept=0.3059322033898326, rvalue=0.6856953093645036, pvalue=0.0, stderr=0.005636250599296884, intercept_stderr=0.00463984862771137)
[[3276 1444]
 [ 512 9412]]

 1 

covariance
[[0.21798569 0.1317351 ]
 [0.1317351  0.18844194]]
correlation
[[1.        0.6499781]
 [0.6499781 1.       ]]
lin regression = LinregressResult(slope=0.604329110263282, intercept=0.337819650067296, rvalue=0.6499780999490244, pvalue=3.5235311898636334e-278, stderr=0.014688579973950545, intercept_stderr=0.012105281579315999)
[[ 492  251]
 [  91 1482]]


## Bivariate analysis by prod_gr_id

In [8]:
for prod_id in sorted(dfcf['prod_gr_id'].unique()):
    df_data_corr = dfcf[dfcf['prod_gr_id'] == prod_id]
    print('\n', prod_id, '\n')
    x, y = df_data_corr['predict_automatch'], df_data_corr['class_acctual']
    print('covariance')
    print(np.cov(x, y))
    print('correlation')
    print(np.corrcoef(x, y))

    result = scipy.stats.linregress(x, y)
    print(f'lin regression = {result}')

    con_matrix = confusion_matrix(x, y)
    print(con_matrix)


 413 

covariance
[[0.22331144 0.11857479]
 [0.11857479 0.18689557]]
correlation
[[1.         0.58041332]
 [0.58041332 1.        ]]
lin regression = LinregressResult(slope=0.5309839460114679, intercept=0.3989713445995599, rvalue=0.580413317348414, pvalue=0.0, stderr=0.0117162421613903, intercept_stderr=0.009543778081147541)
[[ 818  543]
 [ 188 2496]]

 426 

covariance
[[0.21847012 0.14224282]
 [0.14224282 0.19236378]]
correlation
[[1.        0.6938613]
 [0.6938613 1.       ]]
lin regression = LinregressResult(slope=0.6510859127379902, intercept=0.2989214175654847, rvalue=0.6938613039611412, pvalue=0.0, stderr=0.006735632041280073, intercept_stderr=0.00554464717790789)
[[2275  970]
 [ 341 6480]]

 427 

covariance
[[0.21039612 0.15789864]
 [0.15789864 0.19385128]]
correlation
[[1.         0.78185316]
 [0.78185316 1.        ]]
lin regression = LinregressResult(slope=0.7504826775011381, intercept=0.21236872812135232, rvalue=0.7818531571141575, pvalue=0.0, stderr=0.011215849320175813, in

Bivariate analysis by prod_gr_id